pushsync/pusher.go

```diff
@@ -72,27 +77,36 @@ type pushedItem struct {
 // - tags that hold the tags
 func NewPusher(store DB, ps PubSub, tags *chunk.Tags) *Pusher {
 	p := &Pusher{
-		store:    store,
-		tags:     tags,
-		quit:     make(chan struct{}),
-		closed:   make(chan struct{}),
-		pushed:   make(map[string]*pushedItem),
-		receipts: make(chan []byte),
-		ps:       ps,
-		logger:   log.New("self", label(ps.BaseAddr())),
+		store:          store,
+		tags:           tags,
+		quit:           make(chan struct{}),
+		closedChunks:   make(chan struct{}),
+		closedReceipts: make(chan struct{}),
+		pushed:         make(map[string]*pushedItem),
+		receipts:       make(chan []byte),
+		ps:             ps,
+		logger:         log.New("self", label(ps.BaseAddr())),
 	}
-	go p.sync()
+	go p.chunksWorker()
+	go p.receiptsWorker()
 	return p
 }
 
 // Close closes the pusher
 func (p *Pusher) Close() {
 	close(p.quit)
+	timer := time.After(3 * time.Second)
 	select {
-	case <-p.closed:
-	case <-time.After(3 * time.Second):
-		log.Error("timeout closing pusher")
+	case <-p.closedReceipts:
+		select {
+		case <-p.closedChunks:
+			// all closed
+			return
+		case <-timer:
+		}
+	case <-timer:
 	}
+	log.Error("timeout closing pusher")
```

Comment:
?

Reply:
The closing timer is reached at this point.